google/cloud/internal/tracing_rest_client.cc

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/tracing_rest_client.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/internal/rest_opentelemetry.h"
+#include "google/cloud/internal/tracing_http_payload.h"
+#include "google/cloud/internal/tracing_rest_response.h"
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+void ExtractAttributes(StatusOr<std::unique_ptr<RestResponse>> const& value,
+                       opentelemetry::trace::Span& span) {
+  if (!value) return;
+  auto const& response = *value;
+  if (!response) return;
+  for (auto const& kv : response->Headers()) {
+    span.SetAttribute("http.header." + kv.first, kv.second);
+  }
+}
+
+/**
+ * Extracts information from @p value, and adds it to a span.
+ *
+ * The span is ended. The original value is returned, for the sake of
+ * composition.
+ *
+ * Note that this function should be called after the RPC has finished.
+ */
+StatusOr<std::unique_ptr<RestResponse>> EndSpan(
+    opentelemetry::trace::Span& request_span,
+    StatusOr<std::unique_ptr<RestResponse>> value) {
+  if (!value) return internal::EndSpan(request_span, std::move(value));
+  auto payload_span = MakeSpanHttpPayload(request_span);
+  ExtractAttributes(value, *payload_span);
+  value = std::make_unique<TracingRestResponse>(*std::move(value),
+                                                std::move(payload_span));
+  return internal::EndSpan(request_span, std::move(value));
+}
+
+TracingRestClient::TracingRestClient(std::unique_ptr<RestClient> impl)
+    : impl_(std::move(impl)) {}
+
+StatusOr<std::unique_ptr<RestResponse>> TracingRestClient::Delete(
+    RestRequest const& request) {
+  auto span = MakeSpanHttp(request, "DELETE");
+  return EndSpan(*span, impl_->Delete(request));
```

Comment:
>   // Now that I see it, I am not sure we should be linking the payload span to the http verb span.
>  // Also, this is all synchronous code which makes me think we want HttpPayloadSpan to be INTERNAL

The interesting code is something like `storage::Client::ReadObject()`.  Effectively that creates a `RestRequest`, and then saves the `HttpPayload` in an object that may be used long after the `ReadObject()` function returns.  It may be used even in different threads (speaking of that, we should capture the thread ids in these attributes, work for another day).

I think you want to connect the request to its `HttpPayload` object, if they cannot be parent/child, then "link" it is.

Some of the work in `HttpPayload` results in I/O activity, and may trigger RPCs on the server side: as you read data there service will notice this and start sending more data, and triggers RPC.  I do not think `INTERNAL` is the right thing.  But `CONSUMER` may not be it either.  I am changing this to a simple `CLIENT` in #11257. 

In any case. We can fix these things once the overall structure is in place and we look at the UI.

Reply:
FWIW, I am leaving the HttpPayload as its own span, linked to the GetRequest span.

Reply:
> Some of the work in HttpPayload results in I/O activity, and may trigger RPCs on the server side.

Thanks, I did not realize this. I was assuming the `HttpPayload` was a full response.

> `storage::Client::ReadObject()`. Effectively that creates a..

Thanks for summarizing the workflow, but I still do not totally get it. So is there one GET request total? Can/should we send different headers for each `Read` span (from #11257) ?

If you want a second opinion on this stuff, it might be worth chatting/choving. If you know what you are doing and just want someone to stamp PRs, I guess I can do that too.

> FWIW, I am leaving the HttpPayload as its own span, linked to the GetRequest span.

Ack. Span links sounds right to me, btw. Although I do not know whether it should be linked to the GetRequest span, the overall client operation span, or both. It's unfortunate that Cloud Trace is not great with span links, but maybe that is a feature request / bug that can be escalated.

Reply:
> > Some of the work in HttpPayload results in I/O activity, and may trigger RPCs on the server side.
> 
> Thanks, I did not realize this. I was assuming the `HttpPayload` was a full response.
> 
> > `storage::Client::ReadObject()`. Effectively that creates a..
> 
> Thanks for summarizing the workflow, but I still do not totally get it. So is there one GET request total?

Yes.

> Can/should we send different headers for each `Read` span (from #11257) ?

We do not send more data to the service.  We just read the next available data, that (likely) triggers activity in the service.  

> 
> If you want a second opinion on this stuff, it might be worth chatting/choving. If you know what you are doing and just want someone to stamp PRs, I guess I can do that too.

We should talk, because I am not sure what is the "ideal" representation.  I think we should merge this because it is not too terrible and unblocks other improvements.

> 
> > FWIW, I am leaving the HttpPayload as its own span, linked to the GetRequest span.
> 
> Ack. Span links sounds right to me, btw. Although I do not know whether it should be linked to the GetRequest span, the overall client operation span, or both.

It will get even more complicated.  A full download with retry/resume may create multiple GetRequest calls (with their own HTTP/Response spans), in different threads and stack traces, and with the backoffs too.

>  It's unfortunate that Cloud Trace is not great with span links, but maybe that is a feature request / bug that can be escalated.

It is better than I thought:

![image](https://user-images.githubusercontent.com/6241635/231761463-25c3175a-1ae3-45ea-9c4d-3f47f4386fbf.png)